src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/AAsyncCallContext.cs

```diff
@@ -17,38 +17,69 @@ namespace Microsoft.Data.SqlClient
     // CONSIDER creating your own Set method that calls the base Set rather than providing a parameterized ctor, it is friendlier to caching
     // DO NOT use this class' state after Dispose has been called. It will not throw ObjectDisposedException but it will be a cleared object
 
-    internal abstract class AAsyncCallContext<TOwner, TTask> : IDisposable
+    internal abstract class AAsyncCallContext<TOwner, TTask, TDisposable> : AAsyncBaseCallContext<TOwner,TTask>
         where TOwner : class
+        where TDisposable : IDisposable
     {
-        protected TOwner _owner;
-        protected TaskCompletionSource<TTask> _source;
-        protected IDisposable _disposable;
+        protected TDisposable _disposable;
 
         protected AAsyncCallContext()
         {
         }
 
-        protected AAsyncCallContext(TOwner owner, TaskCompletionSource<TTask> source, IDisposable disposable = null)
+        protected AAsyncCallContext(TOwner owner, TaskCompletionSource<TTask> source, TDisposable disposable = default)
         {
             Set(owner, source, disposable);
         }
 
-        protected void Set(TOwner owner, TaskCompletionSource<TTask> source, IDisposable disposable = null)
+        protected void Set(TOwner owner, TaskCompletionSource<TTask> source, TDisposable disposable = default)
+        {
+            base.Set(owner, source);
+            _disposable = disposable;
+        }
+
+        protected override void DisposeCore()
+        {
+            TDisposable copyDisposable = _disposable;
+            _disposable = default;
+            _isDisposed = true;
```

Comment:
isDisposed is already set by parent class in ClearCore()
ReadAsyncCallContext class is another descendant which seems to do it in more consistent way, it does not touch _isDisposed and relies on parent class instead. Maybe isDisposed should be made private..